File: backend/src/services/nameCorruption.ts
/**
 * Broken-font name-corruption detector.
 *
 * Some supplier PDFs embed a subset font without a usable ToUnicode CMap, so text
 * extraction interleaves stray Latin letters / digits INSIDE Cyrillic words
 * (e.g. «Радиатор» -> «PР0а0диатор», «кВт» -> «кBт»). Every stray char is plain
 * ASCII, so it is invisible to replacement-char / control-char quality checks and
 * silently poisons matching & training.
 *
 * Signal = a "sandwich": a Latin letter or digit wedged BETWEEN two Cyrillic letters
 * inside one maximal alphanumeric run. Cyrillic х/Х (U+0445/U+0425) are NOT counted
 * as anchors because they are routinely used as the «×» dimension sign (Ду32х25),
 * which would otherwise false-fire. Validated on the full prod corpus in Stage 1:
 * the invoice-level ratio isolates exactly the fully-corrupt invoices, and the
 * Latin-wedge sub-signal had ZERO false positives across 43 invoices.
 *
 * NOTE: keep this logic in lock-step with scripts/name_corruption.py (regression gate).
 * No hardcoded supplier names or words — pure Unicode structure.
 */

/** Invoice-level corruption ratio at/above which the invoice is no longer a silent "success". */
export const NAME_CORRUPTION_RATIO_THRESHOLD = 0.5;

const isDigit = (code: number): boolean => code >= 0x30 && code <= 0x39;
const isLatin = (code: number): boolean =>
  (code >= 0x41 && code <= 0x5a) || (code >= 0x61 && code <= 0x7a);
const isCyrillic = (code: number): boolean => code >= 0x0400 && code <= 0x04ff;
/** Cyrillic that can anchor a sandwich — excludes х/Х (used as the × dimension sign). */
const isCyrAnchor = (code: number): boolean =>
  isCyrillic(code) && code !== 0x0445 && code !== 0x0425;
/** Chars that belong to a maximal "word" run (split on anything else: space, -, ., /, (, …). */
const isRunChar = (code: number): boolean =>
  isDigit(code) || isLatin(code) || isCyrillic(code);

export interface NameCheck {
  /** A Latin letter OR digit is wedged between two Cyrillic anchors (broad signal). */
  sandwich: boolean;
  /** A Latin LETTER is wedged between two Cyrillic anchors (high-precision: ~0 false positives). */
  latWedge: boolean;
}

/** Inspect a single item name for intra-word script corruption. */
export function isNameCorrupted(name: string): NameCheck {
  let sandwich = false;
  let latWedge = false;
  if (!name) return { sandwich, latWedge };

  const len = name.length;
  let i = 0;
  while (i < len) {
    if (!isRunChar(name.charCodeAt(i))) { i++; continue; }
    // maximal run [i, j)
    let j = i;
    while (j < len && isRunChar(name.charCodeAt(j))) j++;

    let firstAnchor = -1;
    let lastAnchor = -1;
    for (let k = i; k < j; k++) {
      if (isCyrAnchor(name.charCodeAt(k))) {
        if (firstAnchor === -1) firstAnchor = k;
        lastAnchor = k;
      }
    }
    if (firstAnchor !== -1 && lastAnchor > firstAnchor) {
      for (let k = firstAnchor + 1; k < lastAnchor; k++) {
        const code = name.charCodeAt(k);
        if (isDigit(code) || isLatin(code)) {
          sandwich = true;
          if (isLatin(code)) latWedge = true;
        }
      }
    }
    if (sandwich && latWedge) break; // strongest possible verdict reached
    i = j;
  }
  return { sandwich, latWedge };
}

export interface NameCorruptionResult {
  /** flaggedCount / total — invoice-level corruption ratio. */
  ratio: number;
  /** number of names with a sandwich hit. */
  flaggedCount: number;
  total: number;
  /** indices of rows with a high-precision Latin-wedge (safe to flag per-row). */
  latWedgeRows: number[];
}

/** Aggregate corruption over an invoice's item names. */
export function analyzeNameCorruption(names: string[]): NameCorruptionResult {
  const total = names.length;
  if (total === 0) return { ratio: 0, flaggedCount: 0, total: 0, latWedgeRows: [] };

  let flaggedCount = 0;
  const latWedgeRows: number[] = [];
  for (let idx = 0; idx < total; idx++) {
    const r = isNameCorrupted(names[idx]);
    if (r.sandwich) flaggedCount++;
    if (r.latWedge) latWedgeRows.push(idx);
  }
  return { ratio: flaggedCount / total, flaggedCount, total, latWedgeRows };
}

File: scripts/name_corruption.py
"""Broken-font name-corruption detector — PYTHON MIRROR.

Keep this in lock-step with backend/src/services/nameCorruption.ts (identical logic,
identical code points). Used by the regression gate (gate_name_corruption.py).

Signal = a "sandwich": a Latin letter or digit wedged BETWEEN two Cyrillic letters
inside one maximal alphanumeric run. Cyrillic х/Х (U+0445/U+0425) are NOT anchors
(they are used as the «×» dimension sign, e.g. Ду32х25). No hardcoded suppliers/words.
"""

NAME_CORRUPTION_RATIO_THRESHOLD = 0.5


def _is_digit(o): return 0x30 <= o <= 0x39
def _is_latin(o): return (0x41 <= o <= 0x5a) or (0x61 <= o <= 0x7a)
def _is_cyr(o): return 0x0400 <= o <= 0x04ff
def _is_cyr_anchor(o): return _is_cyr(o) and o != 0x0445 and o != 0x0425
def _is_run(o): return _is_digit(o) or _is_latin(o) or _is_cyr(o)


def is_name_corrupted(name):
    """Return (sandwich, lat_wedge) for a single item name."""
    sandwich = False
    lat_wedge = False
    if not name:
        return (sandwich, lat_wedge)
    n = len(name)
    i = 0
    while i < n:
        if not _is_run(ord(name[i])):
            i += 1
            continue
        j = i
        while j < n and _is_run(ord(name[j])):
            j += 1
        first = last = -1
        for k in range(i, j):
            if _is_cyr_anchor(ord(name[k])):
                if first == -1:
                    first = k
                last = k
        if first != -1 and last > first:
            for k in range(first + 1, last):
                o = ord(name[k])
                if _is_digit(o) or _is_latin(o):
                    sandwich = True
                    if _is_latin(o):
                        lat_wedge = True
        if sandwich and lat_wedge:
            break
        i = j
    return (sandwich, lat_wedge)


def analyze_name_corruption(names):
    """Aggregate over an invoice's item names."""
    total = len(names)
    if total == 0:
        return {"ratio": 0.0, "flaggedCount": 0, "total": 0, "latWedgeRows": []}
    flagged = 0
    lat_rows = []
    for idx, nm in enumerate(names):
        s, l = is_name_corrupted(str(nm))
        if s:
            flagged += 1
        if l:
            lat_rows.append(idx)
    return {"ratio": flagged / total, "flaggedCount": flagged,
            "total": total, "latWedgeRows": lat_rows}
